### ISSUE TITLE
Concurrency issues on provision/deprovision

### ISSUE THREAD

**Issue by clebs** [kyma-incubator/compass#823]
**Description**

The provisioner implementation runs both `Provision` and `Deprovision` concurrently in goroutines: https://github.com/kyma-incubator/compass/blob/9b0aa9222afb42f1aab147af84c465a65a553a15/components/provisioner/internal/hydroform/provisioner.go#L61-L67
and
https://github.com/kyma-incubator/compass/blob/9b0aa9222afb42f1aab147af84c465a65a553a15/components/provisioner/internal/hydroform/provisioner.go#L69-L84

Later, the `finish` channel returned by both functions is ignored, which means that the goroutines run uncontrolled and the main goroutine immediately returns:
https://github.com/kyma-incubator/compass/blob/9b0aa9222afb42f1aab147af84c465a65a553a15/components/provisioner/internal/hydroform/provisioner.go#L51-L54
and
https://github.com/kyma-incubator/compass/blob/9b0aa9222afb42f1aab147af84c465a65a553a15/components/provisioner/internal/hydroform/provisioner.go#L56-L59
 
This can cause many different issues and makes it very hard to find and fix bugs (e.g. https://github.com/kyma-incubator/hydroform/issues/61 is probably caused by race conditions)
To describe some dangers:
- Child goroutines can still run after the main goroutine is finished, causing a panic and crashing the whole system.
- Uncontrolled concurrency can easily eat up all the memory, disk and CPU on the system.
- It is possible to run the "same" provision/deprovision call over and over at the same time, causing unpredictable results.
- Results of Hydroform calls (e.g. `ClusterInfo` & `errors`) are accessed without knowing if they are correct, causing wrong conclussions, memory corruption (segfaults) and unpredictable errors.

**Expected result**
Concurrency is managed and and go test/build are run with race detection on.

**Actual result**
See description.

**Comment by Szymongib** [comment]
> Child goroutines can still run after the main goroutine is finished, causing a panic and crashing the whole system.

The main goroutine is handling an HTTP request and it needs to return so that the user gets the response back immediately after successfully (or not) scheduling the Provisioning/Deprovisioning.

> Uncontrolled concurrency can easily eat up all the memory, disk and CPU on the system.

Yes, theoretically we could run some additional goroutine to manage the other goroutines performing Provisioning and Deprovisioning but it still would not be the main goroutine as it is running the HTTP server. Also, the additional goroutines do not run indefinitely they are spawned to do a single Provisioning/Deprovisioning and save the result in the database, after that they just finish. Due to the synchronous nature of Hydroform, we could not cancel the goroutine anyway so I am not really sure how managing those could work.
If you have any ideas I am open to hearing them.

> It is possible to run the "same" provision/deprovision call over and over at the same time, causing unpredictable results.

No, it is not possible. It is blocked by the Application logic. If the Last Operation is InProgress state you cannot start another operation and the operation is set to InProgress before calling the Hydroform.

> Results of Hydroform calls (e.g. ClusterInfo & errors) are accessed without knowing if they are correct, causing wrong conclussions, memory corruption (segfaults) and unpredictable errors.

I am not sure if I understand what do you mean.
Here https://github.com/kyma-incubator/compass/blob/master/components/provisioner/internal/hydroform/service.go#L58 we check if `Provision` returned an error and if that is the case we set Provisioning operation as failed. If there is no error we expect the `ClusterInfo` to not be `nil`. What else could we do?

**Comment by clebs** [comment]
Hi @Szymongib,

generally speaking, when doing concurrency, the threads that are started need to be monitored. The caller always needs to know when a thread will end. Also need to control how many threads are running at the same time. For that, there are many different mechanisms (such as `WaitGroups`).

Specifically to this implementation I see danger in having no control over the running threads, and unfortunately this "In progress" mechanism is not enough to guarantee correct concurrency.

Without proper concurrency control, it is impossible to know what is going to happen, (https://github.com/kyma-incubator/hydroform/issues/61 is a clear example of an unpredictable concurrency issue).

Other than that, I can not really point out any specific problems beforehand because it is, as I said, completely unpredictable. Anything can happen with uncontrolled goroutines, even when one thinks that the application logic is managing it.

My suggestion is to implement a concurrency model that ensures that the code is really doing what we think it is doing.
You can find many different models on the internet (e.g. https://blog.golang.org/pipelines)

**Comment by akgalwas** [comment]
Hi @clebs

By and large, I agree with your statements, however, there is a reasoning behind the current design.

In the main goroutine we start HTTP server. The server handles requests in separate goroutines. Since Hydroform's API is blocking we spin off additional gouroutine that executes Hydroform API methods. As a result:

-  We don't block control flow so that the goroutine created by HTTP server gets finished.
-  Provisioner's client immediately gets response containing operation id that may be later used for checking status.

It has the following consequences:

- Memory consumption : the gouroutines are light weigh but each one calls Hydroform that consumes memory
- Thread safety : if there is some shared state in Hydroform there is a risk of data race

Answering you concerns.

> Specifically to this implementation I see danger in having no control over the running threads, and unfortunately this "In progress" mechanism is not enough to guarantee correct concurrency.

The direct result of current Hydroform API design is:

- the need to make calls in separate goroutines
- the assumption that Hydroform API is thread safe ; if there is a need for any synchronisation it should be done by Hydroform. 
- Doing any synchronization in Provisioner is pointless since it impairs concurrency.

The "In progress" mechanism ensures that we cannot trigged deprovisioning for a cluster that is in provisioning state. So it is impossible to have more than one Hydroform operation per cluster.

> Other than that, I can not really point out any specific problems beforehand because it is, as I said, completely unpredictable. Anything can happen with uncontrolled goroutines, even when one thinks that the application logic is managing it.

From the design perspective goroutines are somewhat similar to Actors in Actor Model. You fire them and don't bother, they do their job. I fully agree that in case we introduce some shared mutable state to the system synchronisation is crucial. In this particular case I cannot see how we could improve the design. We could introduce some kind of manager that would control the gouroutines created, but I have the following concerns:

-  It is impossible to cancel the goroutine once it is started since it performs blocking  call to Hydroform.
- I'm not sure what the manager should do when goroutine finishes.
- I'm not sure how it could synchronise goroutines. Wait group will not work here as the number of goroutines is not known upfront.
- Since it will store bookkeeping data (e.g. channels) it also needs to be thread safe

Theoretically we can come across a situation when main goroutine finishes when there are goroutines running provisioning/deprovisioning. It may happen when HTTP server exits due to failure. In such a case waiting for goroutines to finish would make sense but there is not guarantee that we will exit gracefully before pod is killed. So it seems to me that this is rather corner case and waiting for goroutines would only partially solve the problem. 

The design of Provisioner's code is strongly determined by Hydroform so I cannot see area to improvement.

**Comment by clebs** [comment]
Hi @akgalwas,

the hydroform library is thread safe, with the caveat that it uses the filesystem. But still, each cluster has its exclusive folder where all the file system stuff happens.
Therefore usually it should not have any issues if concurrency is implemented correctly.

On the topic of API design, Hydroform is a library and can be either called concurrently or not, but that always is left up to the caller. That is why the API is designed without any concurrency on it, the caller has to always do that.

This article https://dave.cheney.net/2020/02/23/the-zen-of-go has 2 sections that talk about this topic that I find very relevant:
1. Before you launch a goroutine, know when it will stop
2. Leave concurrency to the caller

To sum up, hydroform is not impeding nor encouraging the use of concurrency, it is completely up to the caller and it is thread safe.
But, if the caller decides to use concurrency, then it needs to be properly implemented. And IMHO, that is not the case right now, which will cause many unpredictable behaviours.

**Comment by akgalwas** [comment]
Hi @clebs,
Hydroform does enforce using concurrency, though not directly. It has blocking nature : single call can take several minutes. So it is not possible to make direct calls to Hydroform API in HTTP handlers since every HTTP call would time out. As it comes to thread safety, I am not sure about Terraform. As far as I understand, the same Terraform code imported in Hydroform is directly used in Terraform CLI. My point is that Terrafrom code is used in different context: Hydroform runs in multithreading context whereas Terraform CLI is single threaded.
In general alternative would be to provide an asynchronous API. Such a design is suitable for long running operations, however, it is more difficult to implement correctly.
As I final thought, I must agree with you that the main issue with the current design in Provisioner is lack of control over go routines creation. It is possible to create many provisioning requests that will result in many go routines running what in turn can crash the Provisioner.

**Comment by janmedrek** [comment]
Handled in: https://github.com/kyma-incubator/compass/issues/1211
